src/Tools/PrepareTests/MinimizeUtil.cs

```diff
@@ -0,0 +1,152 @@
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+
+internal static class MinimizeUtil
+{
+    internal record FilePathInfo(string RelativeDirectory, string Directory, string RelativePath, string FullPath);
+
+    internal static void Run(string sourceDirectory, string destinationDirectory)
+    {
+        // Map of all PE files MVID to the path information
+        var idToFilePathMap = new Dictionary<Guid, List<FilePathInfo>>();
+
+        const string duplicateDirectoryName = ".duplicate";
+        var duplicateDirectory = Path.Combine(destinationDirectory, duplicateDirectoryName);
+        Directory.CreateDirectory(duplicateDirectory);
+
+        InitialWalk();
+        ResolveDuplicates();
+        WriteHydrateFile();
+
+        // The goal of initial walk is to
+        //  1. Record any PE files as they are eligable for de-dup
+        //  2. Hard link all other files into destination directory
+        void InitialWalk()
+        {
+            foreach (var unitDirPath in Directory.EnumerateDirectories(sourceDirectory, "*.UnitTests"))
+            {
+                foreach (var sourceFilePath in Directory.EnumerateFiles(unitDirPath, "*", SearchOption.AllDirectories))
+                {
+                    var currentDirName = Path.GetDirectoryName(sourceFilePath);
+                    var currentRelativeDirectory = Path.GetRelativePath(sourceDirectory, currentDirName!);
+                    var currentOutputDirectory = Path.Combine(destinationDirectory, currentRelativeDirectory);
+                    Directory.CreateDirectory(currentOutputDirectory);
+                    var fileName = Path.GetFileName(sourceFilePath);
+
+                    if (fileName.EndsWith(".dll") && TryGetMvid(sourceFilePath, out var mvid))
+                    {
+                        if (!idToFilePathMap.TryGetValue(mvid, out var list))
+                        {
+                            list = new List<FilePathInfo>();
+                            idToFilePathMap[mvid] = list;
+                        }
+
+                        var filePathInfo = new FilePathInfo(
+                            RelativeDirectory: currentRelativeDirectory,
+                            Directory: currentDirName,
+                            RelativePath: Path.Combine(currentRelativeDirectory, fileName),
+                            FullPath: sourceFilePath);
+                        list.Add(filePathInfo);
+                    }
+                    else
+                    {
+                        var destFilePath = Path.Combine(currentOutputDirectory, fileName);
+                        CreateHardLink(destFilePath, sourceFilePath, IntPtr.Zero);
+                    }
+                }
+            }
+        }
+
+        // Now that we have a complete list of PE files, determine which are duplicates
+        void ResolveDuplicates()
+        {
+            foreach (var pair in idToFilePathMap)
```

Comment:
consider deconstruction: `foreach (var (id, filePaths) in ...`

Reply:
Why spoil all the hackery with beautiful new code? :wink: